src/integer_mod_q/mat_zq/arithmetic/mul.rs

```diff
@@ -51,6 +53,55 @@ impl Mul for &MatZq {
     }
 }
 
+impl Mul<&MatZ> for &MatZq {
+    type Output = MatZq;
+
+    /// Implements the [`Mul`] trait for [`MatZq`] and [`MatZ`].
+    ///
+    /// [`Mul`] is implemented for any combination of owned and borrowed [`MatZq`] and [`MatZ`].
+    ///
+    /// Parameters:
+    /// - `other`: specifies the value to multiply with `self`
+    ///
+    /// Returns the product of `self` and `other` as a [`MatZq`].
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use qfall_math::integer::MatZ;
+    /// use std::str::FromStr;
+    ///
+    /// let a = MatZq::from_str("[[2,1],[1,2]] mod 3").unwrap();
+    /// let b = MatZ::from_str("[[1,0],[0,1]]").unwrap();
+    ///
+    /// let c = &a * &b;
+    /// let d = a * b;
+    /// let e = &c * d;
+    /// let f = c * &e;
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Panics if the dimensions of `self` and `other` do not match for multiplication.
+    fn mul(self, other: &MatZ) -> Self::Output {
+        if self.get_num_columns() != other.get_num_rows() {
+            panic!("Tried to multiply matrices with mismatching matrix dimensions.");
+        }
+
+        let other_matzq = MatZq::from((other, &self.get_mod()));
+        let mut new =
+            MatZq::new(self.get_num_rows(), other.get_num_columns(), self.get_mod()).unwrap();
+        unsafe { fmpz_mod_mat_mul(&mut new.matrix, &self.matrix, &other_matzq.matrix) };
+        new
```

Comment:
This way we do not have to copy the MatZ. Consider adding reduce by a separate function (as I proposed in a different Kanboard task), but if not, I would be fine with my proposal as well.
```suggestion
        let mut new =
            MatZq::new(self.get_num_rows(), other.get_num_columns(), self.get_mod()).unwrap();
        unsafe {
            fmpz_mat_mul(&mut new.matrix.mat[0], &self.matrix.mat[0], &other.matrix);
            _fmpz_mod_mat_reduce(&mut new.matrix)
        }
        new
```